### PR TITLE
Migrate `run-make/issue-14500` to new `rmake.rs` format

### PR DESCRIPTION
**Pull request by Oneirical**
Part of #121876 and the associated [Google Summer of Code project](https://blog.rust-lang.org/2024/05/01/gsoc-2024-selected-projects.html).

Note: I find suspicious that `libbar.a` is hardcoded and is not using the `STATICLIB` call to adapt to Windows platforms. Is this intentional? If not, this will need to be changed.

### CODE REVIEW

File: src/tools/run-make-support/src/rustc.rs [hunk 1]
```diff
@@ -150,6 +150,20 @@ impl Rustc {
         self
     }
 
+    /// Pass a codegen option.
+    pub fn codegen_option(&mut self, option: &str) -> &mut Self {
+        self.cmd.arg("-C");
+        self.cmd.arg(option);
+        self
+    }
```
Comment (jieyouxu) [inline src/tools/run-make-support/src/rustc.rs:158]: I think specific helpers for specific codegen options are better, i.e. `.lto()` over `.codegen_options("lto")` or something.

File: tests/run-make/issue-14500/rmake.rs
Comment (jieyouxu) [inline tests/run-make/issue-14500/rmake.rs:1]: Please take the opportunity to rename this test to something that's not as opaque as `issue-14500`. Maybe `reachable-extern-fn-available-lto`?

File: tests/run-make/issue-14500/rmake.rs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+// is compiled with LTO, it shouldn't strip the symbol from `foo`, and that's the
+// only way that `foo.c` will successfully compile.
+// See https://github.com/rust-lang/rust/issues/14500
+
+//@ ignore-cross-compile
+
+use run_make_support::{cc, extra_c_flags, run, rustc, tmp_dir};
+
+fn main() {
+    let libbar_path = tmp_dir().join("libbar.a");
```
Comment (jieyouxu) [inline tests/run-make/issue-14500/rmake.rs:15]: I don't recall exactly, do we have a static library name helper? If not, can you add one?

Comment (Oneirical) [reply tests/run-make/issue-14500/rmake.rs:15]: There is already one! I should use it.

**Comment by Oneirical** [comment]
> Well noticed. It just so happens that everyone (that we care about) uses just about the same naming convention for static libraries:

Here is the part that made me suspicious (in `tools.mk`):

`mylibrary.lib` on MSVC.

```makefile
     ifeq ($(UNAME),Darwin)
         STATICLIB = $(TMPDIR)/lib$(1).a
     else
         ifdef IS_WINDOWS
             ifdef IS_MSVC
                 STATICLIB = $(TMPDIR)/$(1).lib
             else
                 STATICLIB = $(TMPDIR)/lib$(1).a
             endif
         else
             STATICLIB = $(TMPDIR)/lib$(1).a
         endif
     endif
 ```

**Comment by jieyouxu** [comment]
> > Well noticed. It just so happens that everyone (that we care about) uses just about the same naming convention for static libraries:
> 
> Here is the part that made me suspicious (in `tools.mk`):
> 
> `mylibrary.lib` on MSVC.
> 
> ```makefile
>      ifeq ($(UNAME),Darwin)
>          STATICLIB = $(TMPDIR)/lib$(1).a
>      else
>          ifdef IS_WINDOWS
>              ifdef IS_MSVC
>                  STATICLIB = $(TMPDIR)/$(1).lib
>              else
>                  STATICLIB = $(TMPDIR)/lib$(1).a
>              endif
>          else
>              STATICLIB = $(TMPDIR)/lib$(1).a
>          endif
>      endif
> ```

You're right, I completely forgot about msvc being special again... There's technically no standard here, but AFAICT msvc uses `lib` by its own convention. So yeah, we can use the `static_library_name` helper in the support library.

**Comment by jieyouxu** [comment]
Thanks!

@bors r+ rollup

**Pull request merged**